examples/generative/finetune_stable_diffusion.py

```diff
@@ -0,0 +1,438 @@
+    return augmenter(image_batch), token_batch
+
+
+def run_text_encoder(image_batch, token_batch):
+    return (
+        image_batch,
+        token_batch,
+        text_encoder([token_batch, POS_IDS], training=False),
+    )
+
+
+def prepare_dict(image_batch, token_batch, encoded_text_batch):
+    return {
+        "images": image_batch,
+        "tokens": token_batch,
+        "encoded_text": encoded_text_batch,
+    }
+
+
+def prepare_dataset(image_paths, tokenized_texts, batch_size=1):
+    dataset = tf.data.Dataset.from_tensor_slices((image_paths, tokenized_texts))
+    dataset = dataset.shuffle(batch_size * 10)
+    dataset = dataset.map(process_image, num_parallel_calls=AUTO).batch(batch_size)
+    dataset = dataset.map(apply_augmentation, num_parallel_calls=AUTO)
+    dataset = dataset.map(run_text_encoder, num_parallel_calls=AUTO)
+    dataset = dataset.map(prepare_dict, num_parallel_calls=AUTO)
+    return dataset.prefetch(AUTO)
+
+
+"""
+The baseline Stable Diffusion model was trained using images of 512x512 resolution. It's
+unlikely for a model that's trained using higher resolution imaged to transfer well to
+lower resolution images. However, the current model will lead to OOM if we keep the
+resolution to 512x512 (without enabling mixed-precision). Therefore, in the interest of
+interactive demonstration, we kept the input resolution to 256x256.
+"""
+
+# Prepare the dataset.
+training_dataset = prepare_dataset(
+    np.array(data_frame["image_path"]), tokenized_texts, batch_size=4
+)
+
+# Take a sample batch and investigate.
+sample_batch = next(iter(training_dataset))
+
+for k in sample_batch:
+    print(k, sample_batch[k].shape)
+
+"""
+We can also take a look at the training images and their corresponding captions.
+"""
+
+plt.figure(figsize=(20, 10))
+
+for i in range(3):
+    ax = plt.subplot(1, 4, i + 1)
+    plt.imshow((sample_batch["images"][i] + 1) / 2)
+
+    text = tokenizer.decode(sample_batch["tokens"][i].numpy().squeeze())
+    text = text.replace("<|startoftext|>", "")
+    text = text.replace("<|endoftext|>", "")
+    plt.title(text)
+
+    plt.axis("off")
+
+"""
+## A trainer class for the fine-tuning loop
+"""
+
+
+class Trainer(tf.keras.Model):
+    # Reference:
+    # https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py
+
+    def __init__(
+        self, diffusion_model, vae, noise_scheduler, max_grad_norm=1.0, **kwargs
+    ):
+        super().__init__(**kwargs)
+
+        self.diffusion_model = diffusion_model
+        self.vae = vae
+        self.noise_scheduler = noise_scheduler
+        self.max_grad_norm = max_grad_norm
+
+        self.vae.trainable = False
+
+    def train_step(self, inputs):
+        images = inputs["images"]
+        encoded_text = inputs["encoded_text"]
+        bsz = tf.shape(images)[0]
+
+        with tf.GradientTape() as tape:
+            # Project image into the latent space.
+            latents = self.vae(images, training=False)
```

Comment:
We should ideally sample from the variance and mean predicted by the VAE instead of taking its final outputs directly. We're currently experimenting with it ([code branch](https://github.com/sayakpaul/stable-diffusion-keras-ft/tree/feat/sampler)) and will provide an update as soon as we have some results. But I guess it doesn't block the reviewing. 

Cc: @deep-diver

Reply:
This has been addressed. 

Since we're training the model from scratch, it'll take some time to get to an epoch checkpoint producing plausible image quality. Once we have the results, we'll update the PR with those.